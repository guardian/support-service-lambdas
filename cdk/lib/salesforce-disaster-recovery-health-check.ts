import { GuScheduledLambda } from '@guardian/cdk';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import type { App } from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { Schedule } from 'aws-cdk-lib/aws-events';
import { PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { StateMachine } from 'aws-cdk-lib/aws-stepfunctions';

export class SalesforceDisasterRecoveryHealthCheck extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const app = 'salesforce-disaster-recovery-health-check';

		const stateMachine = StateMachine.fromStateMachineName(
			this,
			'sdf',
			`salesforce-disaster-recovery-${this.stage}`,
		);

		new GuScheduledLambda(this, 'salesforce-disaster-recovery-health-check', {
			app,
			memorySize: 1024,
			fileName: `${app}.zip`,
			runtime: Runtime.NODEJS_20_X,
			timeout: Duration.minutes(5),
			environment: {
				APP: app,
				STACK: this.stack,
				STAGE: this.stage,
				STATE_MACHINE_ARN: stateMachine.stateMachineArn,
			},
			handler: 'salesforceDisasterRecoveryHealthCheck.handler',
			functionName: `${app}-${this.stage}`,
			rules: [{ schedule: Schedule.cron({ minute: '0', hour: '9' }) }],
			monitoringConfiguration: { noMonitoring: true },
			initialPolicy: [
				new PolicyStatement({
					actions: ['states:StartExecution', 'states:DescribeExecution'],
					resources: [stateMachine.stateMachineArn],
				}),
			],
		});
	}
}

// lambdaFunction.addToRolePolicy(new iam.PolicyStatement({
// 	actions: ['states:StartExecution', 'states:DescribeExecution'],
// 	resources: ['arn:aws:states:REGION:ACCOUNT_ID:stateMachine:YOUR_STATE_MACHINE_NAME'] // replace placeholders
//   }));
//   topic.grantPublish(lambdaFunction);
