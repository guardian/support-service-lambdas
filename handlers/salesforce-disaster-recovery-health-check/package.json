{
	"name": "salesforce-disaster-recovery-health-check",
	"scripts": {
		"test": "jest --group=-integration",
		"it-test": "jest --group=integration",
		"type-check": "tsc --noEmit",
		"build": "esbuild --bundle --platform=node --target=node20 --outdir=target src/handlers/*.ts",
		"lint": "eslint src/**/*.ts",
		"package": "pnpm type-check && pnpm lint && pnpm check-formatting && pnpm test && pnpm build && cd target && zip -qr salesforce-disaster-recovery-health-check.zip ./*.js",
		"check-formatting": "prettier --check **/*.ts",
		"fix-formatting": "prettier --write **/*.ts"
	},
	"devDependencies": {
		"@types/aws-lambda": "^8.10.142"
	},
	"dependencies": {
		"@aws-sdk/client-sfn": "^3.552.0",
		"@aws-sdk/client-sns": "^3.552.0"
	}
}
