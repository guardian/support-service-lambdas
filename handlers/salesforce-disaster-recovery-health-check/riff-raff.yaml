stacks:
  - support
regions:
  - eu-west-1
allowedStages:
  - CODE
  - PROD
deployments:
  salesforce-disaster-recovery-health-check-cloudformation:
    type: cloud-formation
    app: salesforce-disaster-recovery-health-check
    parameters:
      templateStagePaths:
        CODE: salesforce-disaster-recovery-health-check-CODE.template.json
        PROD: salesforce-disaster-recovery-health-check-PROD.template.json

  salesforce-disaster-recovery-health-check:
    type: aws-lambda
    parameters:
      fileName: salesforce-disaster-recovery-health-check.zip
      bucketSsmLookup: true
      prefixStack: false
      functionNames:
        - salesforce-disaster-recovery-health-check-
    dependencies: [salesforce-disaster-recovery-health-check-cloudformation]
