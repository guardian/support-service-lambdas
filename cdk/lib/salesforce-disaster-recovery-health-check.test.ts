import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { SalesforceDisasterRecoveryHealthCheck } from './salesforce-disaster-recovery-health-check';

describe('The Salesforce disaster recovery health check stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const codeStack = new SalesforceDisasterRecoveryHealthCheck(
			app,
			'salesforce-disaster-recovery-health-check-CODE',
			{
				stack: 'membership',
				stage: 'CODE',
			},
		);
		const prodStack = new SalesforceDisasterRecoveryHealthCheck(
			app,
			'salesforce-disaster-recovery-health-check-PROD',
			{
				stack: 'membership',
				stage: 'PROD',
			},
		);

		expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
		expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
	});
});
