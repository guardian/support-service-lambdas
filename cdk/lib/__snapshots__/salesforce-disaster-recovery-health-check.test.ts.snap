// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Salesforce disaster recovery health check stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "salesforcedisasterrecoveryhealthcheck7A2C7C8E": {
      "DependsOn": [
        "salesforcedisasterrecoveryhealthcheckServiceRoleDefaultPolicy0ACEC506",
        "salesforcedisasterrecoveryhealthcheckServiceRoleAAFD4114",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "membership/CODE/salesforce-disaster-recovery-health-check/salesforce-disaster-recovery-health-check.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "salesforce-disaster-recovery-health-check",
            "STACK": "membership",
            "STAGE": "CODE",
            "STATE_MACHINE_ARN": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":states:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":stateMachine:salesforce-disaster-recovery-CODE",
                ],
              ],
            },
          },
        },
        "FunctionName": "salesforce-disaster-recovery-health-check-CODE",
        "Handler": "salesforceDisasterRecoveryHealthCheck.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "salesforcedisasterrecoveryhealthcheckServiceRoleAAFD4114",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "salesforce-disaster-recovery-health-check",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "salesforcedisasterrecoveryhealthcheckServiceRoleAAFD4114": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "salesforce-disaster-recovery-health-check",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "salesforcedisasterrecoveryhealthcheckServiceRoleDefaultPolicy0ACEC506": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:states:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":stateMachine:salesforce-disaster-recovery-CODE",
                  ],
                ],
              },
            },
            {
              "Action": "states:DescribeExecution",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:states:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":execution:salesforce-disaster-recovery-CODE:*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/membership/CODE/salesforce-disaster-recovery-health-check/salesforce-disaster-recovery-health-check.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/membership/salesforce-disaster-recovery-health-check",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/membership/salesforce-disaster-recovery-health-check/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "salesforcedisasterrecoveryhealthcheckServiceRoleDefaultPolicy0ACEC506",
        "Roles": [
          {
            "Ref": "salesforcedisasterrecoveryhealthcheckServiceRoleAAFD4114",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "salesforcedisasterrecoveryhealthchecksalesforcedisasterrecoveryhealthcheckcron06MON05DCAE5E4": {
      "Properties": {
        "ScheduleExpression": "cron(0 6 ? * MON *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "salesforcedisasterrecoveryhealthcheck7A2C7C8E",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "salesforcedisasterrecoveryhealthchecksalesforcedisasterrecoveryhealthcheckcron06MON0AllowEventRulesalesforcedisasterrecoveryhealthcheckCODEsalesforcedisasterrecoveryhealthcheck1E72AB495CD3090A": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "salesforcedisasterrecoveryhealthcheck7A2C7C8E",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "salesforcedisasterrecoveryhealthchecksalesforcedisasterrecoveryhealthcheckcron06MON05DCAE5E4",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;

exports[`The Salesforce disaster recovery health check stack matches the snapshot 2`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "salesforcedisasterrecoveryhealthcheck7A2C7C8E": {
      "DependsOn": [
        "salesforcedisasterrecoveryhealthcheckServiceRoleDefaultPolicy0ACEC506",
        "salesforcedisasterrecoveryhealthcheckServiceRoleAAFD4114",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "membership/PROD/salesforce-disaster-recovery-health-check/salesforce-disaster-recovery-health-check.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "salesforce-disaster-recovery-health-check",
            "STACK": "membership",
            "STAGE": "PROD",
            "STATE_MACHINE_ARN": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":states:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":stateMachine:salesforce-disaster-recovery-PROD",
                ],
              ],
            },
          },
        },
        "FunctionName": "salesforce-disaster-recovery-health-check-PROD",
        "Handler": "salesforceDisasterRecoveryHealthCheck.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "salesforcedisasterrecoveryhealthcheckServiceRoleAAFD4114",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "salesforce-disaster-recovery-health-check",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "salesforcedisasterrecoveryhealthcheckServiceRoleAAFD4114": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "salesforce-disaster-recovery-health-check",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "salesforcedisasterrecoveryhealthcheckServiceRoleDefaultPolicy0ACEC506": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:states:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":stateMachine:salesforce-disaster-recovery-PROD",
                  ],
                ],
              },
            },
            {
              "Action": "states:DescribeExecution",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:states:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":execution:salesforce-disaster-recovery-PROD:*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/membership/PROD/salesforce-disaster-recovery-health-check/salesforce-disaster-recovery-health-check.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/membership/salesforce-disaster-recovery-health-check",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/membership/salesforce-disaster-recovery-health-check/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "salesforcedisasterrecoveryhealthcheckServiceRoleDefaultPolicy0ACEC506",
        "Roles": [
          {
            "Ref": "salesforcedisasterrecoveryhealthcheckServiceRoleAAFD4114",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "salesforcedisasterrecoveryhealthchecksalesforcedisasterrecoveryhealthcheckcron06MON05DCAE5E4": {
      "Properties": {
        "ScheduleExpression": "cron(0 6 ? * MON *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "salesforcedisasterrecoveryhealthcheck7A2C7C8E",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "salesforcedisasterrecoveryhealthchecksalesforcedisasterrecoveryhealthcheckcron06MON0AllowEventRulesalesforcedisasterrecoveryhealthcheckPRODsalesforcedisasterrecoveryhealthcheck330F201C9F7FB4EF": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "salesforcedisasterrecoveryhealthcheck7A2C7C8E",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "salesforcedisasterrecoveryhealthchecksalesforcedisasterrecoveryhealthcheckcron06MON05DCAE5E4",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
